### PR TITLE
Trying to remove blink during review

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
I follow the idea given in https://github.com/ankidroid/Anki-Android/pull/6224 . Given the feedback I received, I don't expect this PR to be accepted, but it was quick to do and to test, so I thought I'd still give it a try.

There are three ways to block buttons. Not blocking (i.e. buttons are
activated), quick blocking and slow blocking. Buttons are grayed only
if we expect the blocking to be slow. I.e. when we need to recompute
the queue because the user decided to undo/bury/suspend or just loaded
the deck. Otherwise, we expect the next card to load quickly, and so
can block the buttons without graying them. This will ensure that the
buttons does not blink.

Note that loading the next card can sometime be slow, so it's still
not a perfect solution.

## Purpose / Description
Users sees buttons blinking/flickering when they review. This is because buttons are deactivated and reactivated immediately.

## Fixes
Only change color if we expect the scheduler to take time to give the new card, because we know that it needs to recompute the queue. Otherwise, assume that the card will come quickly and so changing the color is useless.

## Approach
Avoiding flicker for the most common case. Letting potential flickr for the slower and rarer actions.

## How Has This Been Tested?

Android emulator, with a few card reviewed and cancelled.

## Learning (optional, can help others)
I did the research when I tried to find where ankidroid was slow. I have no made any research for this bug right now. The idea actually arrived when I read "TL;DR: Predicting future events is hard" in https://github.com/ankidroid/Anki-Android/pull/6224#issuecomment-631074449 and realized that I was disagreeing with it, and that I already have a model of which events are quick/slow

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
Note that https://github.com/ankidroid/Anki-Android/pull/6227/commits/2efd608724403092dba60839bec11436476152b9 is the commit which really correct this error. The two previous commits are simply here to clean the code, and the last commit deal with an unrelated problem: the undo button was not cancelled

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -2244,7 +2243,7 @@ private boolean typeAnswer() {
 
 
     private void unblockControls() {
-        mControlBlocked = false;
+        mControlBlocked = ControlBlock.NONE;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:2246]: nit: `UNBLOCKED` would read better than `NONE`

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 2]
```diff
@@ -786,7 +784,7 @@ public void onWindowFocusChanged(boolean hasFocus) {
      * @return true if there is another card of same note that could be dismissed
      */
     private boolean suspendNoteAvailable() {
-        if (mCurrentCard == null || getControlBlocked()) {
+        if (mCurrentCard == null || getControlBlocked() != ControlBlock.NONE) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:787]: `getControlBlocked() != ControlBlock.NONE` is duplicated, could it be extracted into a method

File: AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ActionButtonStatus.java [hunk 3]
```diff
@@ -73,16 +73,17 @@ public void setCustomButtons(Menu menu) {
                 MenuItem item = menu.findItem(itemId);
                 item.setShowAsAction(mCustomButtons.get(itemId));
                 Drawable icon = item.getIcon();
-                if (mReviewerUi.getControlBlocked()) {
-                    item.setEnabled(false);
-                    if (icon != null) {
+                if (icon != null) {
+                    if (mReviewerUi.getControlBlocked() == ReviewerUi.ControlBlock.SLOW) {
                         icon.setAlpha(Themes.ALPHA_ICON_DISABLED_LIGHT);
+                    } else {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ActionButtonStatus.java:79]: Could a short comment be added explaining the intention

**Comment by mikehardy** [comment]
> Nothing major, simple and easy to read. Main note: [ddd80f1](https://github.com/ankidroid/Anki-Android/commit/ddd80f11f1482e45b4436b6d18551f3db61f6cad) was reverted due to #6119.
> 
> Is there a compromise that we can take?

This the most important feedback as it did result in an issue and subsequent PR

I could go this route though, code-wise more complicated than just toasting but UI-wise toasting after a user tries something is a bit of a UI failure (the user should have known via consistent UI state, which this would implement)

**Comment by Arthur-Milchior** [comment]
Ill remove the undo commit and put it in another PR then. No need to consider both together

**Comment by Arthur-Milchior** [comment]
@david-allison-1 's feed back are taken into account in this new version

File: AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ActionButtonStatus.java [hunk 4]
```diff
@@ -73,14 +73,23 @@ public void setCustomButtons(Menu menu) {
...
                 } else {
                     item.setEnabled(true);
-                    if (icon != null) {
+                }
+                if (icon != null) {
+                    /* Ideially, we want to give feedback to users that buttons are disabled.
+                    However, some actions are expected to be so quick that the visual feedback is useless
+                    and is only seen as a flickering.
+
+                    We use an heuristic to decide whether the next card will appear quickly or slowly.
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ActionButtonStatus.java:86]:
```suggestion
                    We use a heuristic to decide whether the next card will appear quickly or slowly.
```

File: AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ReviewerUi.java [hunk 5]
```diff
@@ -17,5 +17,19 @@
 package com.ichi2.anki.reviewer;
 
 public interface ReviewerUi {
-    boolean getControlBlocked();
+    /** How to block UI buttons. */
+    enum ControlBlock {
+        /** Buttons are functionals*/
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ReviewerUi.java:22]:
```suggestion
        /** Buttons are functional */
```

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerKeyboardInputTest.java [hunk 6]
```diff
@@ -267,15 +268,20 @@ private void assertGamepadButtonAnswers(int keycodeButton, int ease) {
...
 
         @Override
-        public boolean getControlBlocked() {
+        public ControlBlock getControlBlocked() {
             return mControlsAreBlocked;
         }
 
+        @Override
+        public boolean isControlBlocked() {
+            return getControlBlocked() != ControlBlock.UNBLOCKED;
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/ReviewerKeyboardInputTest.java:282]: I don't think this needs to exist as it's the same as the base class

File: AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ReviewerUi.java (hunk 5 above)
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ReviewerUi.java:22]:
```suggestion
        /** Buttons are non-functional */
```

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ReviewerUi.java:22]: Nope. If the buttons are unblocked, then they are functional.

**Comment by Arthur-Milchior** [comment]
Minor corrections done.
I split the code in smaller commit for readability

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ActionButtonStatus.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5555198&bid=18217880)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ AnkiDroid/src/main/java/com/ichi2/anki/reviewer/ActionButtonStatus.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5555198&bid=18217880)

**Comment by mikehardy** [comment]
Going out now as 2.10.3beta3 and 2.11alpha6

**Comment by Arthur-Milchior** [comment]
At this point, I officially don't even understand the version numbers.

**Comment by david-allison-1** [comment]
> At this point, I officially don't even understand the version numbers.

2.10.3beta3 - This is the third beta of 2.10.3. When stable, it will be 2.10.3  
2.11alpha6. This is the sixth alpha of 2.11. When stable, it will be 2.11

**Comment by mikehardy** [comment]
Yeah, parallel release tracks to manage different levels of code stability. We branch for public and beta releases (current branch is release-2.10 that all the 2.10 releases are coming from) but for alpha releases they go right from master. For bugfixes or things self-evidently stable I cherry-pick them from git master branch to the release branch https://github.com/ankidroid/Anki-Android/wiki/Release-procedure

**Comment by Arthur-Milchior** [comment]
I believe I start to get it. I was expecting that the 11alpha track contained everything that was in 2.10. So that this would be redundant. I now realize that you need to had everything into .11alpha too because you don't merge 10 in 11

**Comment by mikehardy** [comment]
Exactly. Everything goes in master first. Release branches are branched off for public release and they will eventually die as master keeps on moving. Getting fixes into public releases is a one-way process of `git cherry-pick -x commit-hash` from master to the release branch

**Pull request merged**